spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -404,6 +404,53 @@ contract RolloverVault is
         return notes;
     }
 
+    /// @notice Deposits the underlying asset from {msg.sender} into the perp and vault systems,
+    ///         and mints both perp tokens and notes.
+    /// @param amount The amount tokens to be deposited into the vault.
+    /// @return The amount of perps and vault notes minted.
+    function deposit2(uint256 amount) external nonReentrant whenNotPaused returns (uint256, uint256) {
+        if (amount <= 0) {
+            revert UnacceptableDeposit();
+        }
+
+        uint256 totalSupply_ = totalSupply();
+        uint256 tvlBefore = getTVL();
+
+        // get deposit bond
+        IBondController depositBond = perp.getDepositBond();
+        BondTranches memory bt = depositBond.getTranches();
+
+        // NOTE: the vault may already have seniors which
+        //       we have to ignore for the sake of this operation.
+        uint256 seniorBalBefore = bt.tranches[0].balanceOf(address(this));
+
+        // transfer underlying in
+        underlying.safeTransferFrom(_msgSender(), address(this), amount);
+
+        // tranches underlying and updates vault assets with fresh tranches
+        _tranche(depositBond, bt, amount);
+
+        // mint perps using freshly tranched seniors
+        uint256 seniorsMinted = bt.tranches[0].balanceOf(address(this)) - seniorBalBefore;
```

Comment:
It seems philosophically safer to not ever interact with the vaults existing tranche balances at all. Until the final upkeep, anyway, once the main operations are complete. This closes down a whole class of bugs & attacks.

When we tranche the underlying, shouldn't that function be able to return the number of tranches created? If that means not using the `_tranche` helper, that seems ok to me. That helper also updates the vault assets with intermediate values that we don't want anyway.

Reply:
In the current version, the tranche function does not return the number of tranches created. We'll have to figure that out based on the balance changes ..

https://github.com/buttonwood-protocol/tranche/blob/main/contracts/BondController.sol#L123

Reply:
Ok, in that case let's stick with your original implementation